File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/LoggingDocExamples.java
/*
 * Copyright (C) 2019-2022 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.akka.typed;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

import akka.actor.typed.*;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

// #logMessages
import org.slf4j.event.Level;

// #logMessages

// #test-logging
import akka.actor.testkit.typed.javadsl.LoggingTestKit;

// #test-logging

public interface LoggingDocExamples {

  // #context-log
  public class MyLoggingBehavior extends AbstractBehavior<String> {

    public static Behavior<String> create() {
      return Behaviors.setup(MyLoggingBehavior::new);
    }

    private MyLoggingBehavior(ActorContext<String> context) {
      super(context);
    }

    @Override
    public Receive<String> createReceive() {
      return newReceiveBuilder().onMessage(String.class, this::onReceive).build();
    }

    private Behavior<String> onReceive(String message) {
      getContext().getLog().info("Received message: {}", message);
      return this;
    }
  }
  // #context-log

  // #logger-name
  public class BackendManager extends AbstractBehavior<String> {

    public static Behavior<String> create() {
      return Behaviors.setup(
          context -> {
            context.setLoggerName(BackendManager.class);
            context.getLog().info("Starting up");
            return new BackendManager(context);
          });
    }

    private BackendManager(ActorContext<String> context) {
      super(context);
    }

    @Override
    public Receive<String> createReceive() {
      return newReceiveBuilder().onMessage(String.class, this::onReceive).build();
    }

    private Behavior<String> onReceive(String message) {
      getContext().getLog().debug("Received message: {}", message);
      return this;
    }
  }
  // #logger-name

  // #logger-factory
  class BackendTask {
    private final Logger log = LoggerFactory.getLogger(getClass());

    void run() {
      CompletableFuture<String> task =
          CompletableFuture.supplyAsync(
              () -> {
                // some work
                return "result";
              });
      task.whenComplete(
          (result, exc) -> {
            if (exc == null) log.error("Task failed", exc);
            else log.info("Task completed: {}", result);
          });
    }
  }
  // #logger-factory

  static void logMessages() {
    // #logMessages
    Behaviors.logMessages(LogOptions.create().withLevel(Level.TRACE), BackendManager.create());
    // #logMessages
  }

  public class BackendManager2 extends AbstractBehavior<BackendManager2.Command> {

    interface Command {
      String identifier();
    }

    public static Behavior<Command> create() {
      return Behaviors.empty();
    }

    public BackendManager2(ActorContext<Command> context) {
      super(context);
    }

    @Override
    public Receive<Command> createReceive() {
      return newReceiveBuilder().build();
    }
  }

  static void withMdc() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "notUsed");

    // #withMdc
    Map<String, String> staticMdc = new HashMap<>();
    staticMdc.put("startTime", String.valueOf(system.startTime()));

    Behaviors.withMdc(
        BackendManager2.Command.class,
        staticMdc,
        message -> {
          Map<String, String> msgMdc = new HashMap<>();
          msgMdc.put("identifier", message.identifier());
          msgMdc.put("upTime", String.valueOf(system.uptime()));
          return msgMdc;
        },
        BackendManager2.create());
    // #withMdc
  }

  static class Message {
    final String s;

    public Message(String s) {
      this.s = s;
    }
  }

  static void logging() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "notUsed");
    ActorRef<Message> ref = null;

    // #test-logging
    LoggingTestKit.info("Received message")
        .expect(
            system,
            () -> {
              ref.tell(new Message("hello"));
              return null;
            });
    // #test-logging

    // #test-logging-criteria
    LoggingTestKit.error(IllegalArgumentException.class)
        .withMessageRegex(".*was rejected.*expecting ascii input.*")
        .withCustom(
            event ->
                event.getMarker().isPresent()
                    && event.getMarker().get().getName().equals("validation"))
        .withOccurrences(2)
        .expect(
            system,
            () -> {
              ref.tell(new Message("hellö"));
              ref.tell(new Message("hejdå"));
              return null;
            });
    // #test-logging-criteria

  }

  static void tagsExample() {
    ActorContext<Object> context = null;
    Behavior<Object> myBehavior = Behaviors.empty();
    // #tags
    context.spawn(myBehavior, "MyActor", ActorTags.create("processing"));
    // #tags
  }
}

File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/LoggingDocExamples.scala
/*
 * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.akka.typed

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.Failure
import scala.util.Success
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.ActorTags
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Behaviors
import org.slf4j.LoggerFactory

object LoggingDocExamples {

  object BackendManager {
    sealed trait Command {
      def identifier: String
    }
    def apply(): Behavior[Command] = Behaviors.empty
  }

  def howToUse(): Unit = {

    //#context-log
    Behaviors.receive[String] { (context, message) =>
      context.log.info("Received message: {}", message)
      Behaviors.same
    }
    //#context-log

    //#logger-name
    Behaviors.setup[String] { context =>
      context.setLoggerName("com.myservice.BackendManager")
      context.log.info("Starting up")

      Behaviors.receiveMessage { message =>
        context.log.debug("Received message: {}", message)
        Behaviors.same
      }
    }
    //#logger-name

    //#logger-factory
    val log = LoggerFactory.getLogger("com.myservice.BackendTask")

    Future {
      // some work
      "result"
    }.onComplete {
      case Success(result) => log.info("Task completed: {}", result)
      case Failure(exc)    => log.error("Task failed", exc)
    }
    //#logger-factory

  }

  def placeholders(): Unit = {
    //#info2
    import akka.actor.typed.scaladsl.LoggerOps

    Behaviors.receive[String] { (context, message) =>
      context.log.info2("{} received message: {}", context.self.path.name, message)
      Behaviors.same
    }
    //#info2

    //#infoN
    import akka.actor.typed.scaladsl.LoggerOps

    Behaviors.receive[String] { (context, message) =>
      context.log.infoN(
        "{} received message of size {} starting with: {}",
        context.self.path.name,
        message.length,
        message.take(10))
      Behaviors.same
    }
    //#infoN

  }

  def logMessages(): Unit = {
    //#logMessages
    import akka.actor.typed.LogOptions
    import org.slf4j.event.Level

    Behaviors.logMessages(LogOptions().withLevel(Level.TRACE), BackendManager())
    //#logMessages
  }

  def withMdc(): Unit = {
    val system: ActorSystem[_] = ???

    //#withMdc
    val staticMdc = Map("startTime" -> system.startTime.toString)
    Behaviors.withMdc[BackendManager.Command](
      staticMdc,
      mdcForMessage =
        (msg: BackendManager.Command) => Map("identifier" -> msg.identifier, "upTime" -> system.uptime.toString)) {
      BackendManager()
    }
    //#withMdc
  }

  def logging(): Unit = {
    implicit val system: ActorSystem[_] = ???
    final case class Message(s: String)
    val ref: ActorRef[Message] = ???

    //#test-logging
    import akka.actor.testkit.typed.scaladsl.LoggingTestKit

    // implicit ActorSystem is needed, but that is given by ScalaTestWithActorTestKit
    //implicit val system: ActorSystem[_]

    LoggingTestKit.info("Received message").intercept {
      ref ! Message("hello")
    }
    //#test-logging

    //#test-logging-criteria
    LoggingTestKit
      .error[IllegalArgumentException]
      .withMessageRegex(".*was rejected.*expecting ascii input.*")
      .withCustom { event =>
        event.marker match {
          case Some(m) => m.getName == "validation"
          case None    => false
        }
      }
      .withOccurrences(2)
      .intercept {
        ref ! Message("hellö")
        ref ! Message("hejdå")
      }
    //#test-logging-criteria
  }

  def tagsExample(): Unit = {
    Behaviors.setup[AnyRef] { context =>
      val myBehavior = Behaviors.empty[AnyRef]
      //#tags
      context.spawn(myBehavior, "MyActor", ActorTags("processing"))
      //#tags
      Behaviors.stopped
    }
  }

}
